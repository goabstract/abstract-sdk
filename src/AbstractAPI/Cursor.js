/* @flow */
import type { CursorMeta, CursorPromise, CursorResponse } from "../";

export default class Cursor<T> {
  lastResponse: ?Promise<CursorResponse<T> | void>;
  meta: CursorMeta;
  request: (meta?: CursorMeta) => Promise<CursorResponse<T>>;

  _next(first?: boolean): CursorPromise<T> {
    const makeRequest = async () => {
      if (!first && !this.meta.nextOffset) return;
      const response = await this.request(this.meta);
      this.meta = response.meta;
      return response;
    };
    this.lastResponse = this.lastResponse
      ? this.lastResponse.then(makeRequest)
      : makeRequest();
    const promise = (this.lastResponse.then(
      response => response && response.data
    ): any);
    promise.next = this.next.bind(this);
    const cursorPromise = (promise: CursorPromise<T>);
    return cursorPromise;
  }

  constructor(
    request: (meta?: CursorMeta) => Promise<CursorResponse<T>>
  ): CursorPromise<T> {
    this.request = request;
    return ((this: any): CursorPromise<T>);
  }

  then(
    onFulfilled: (data: T) => *,
    onRejected: (error: *) => *
  ): CursorPromise<T> {
    const cursorPromise = this._next(true);
    cursorPromise.then(onFulfilled, onRejected);
    return cursorPromise;
  }

  next(): CursorPromise<T> {
    return this._next();
  }
}
