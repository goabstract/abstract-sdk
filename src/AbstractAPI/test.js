// @flow
import fs from "fs";
import path from "path";
import fetch from "jest-fetch-mock";
import get from "lodash/get";
import {
  buildOptions,
  buildOrganizationDescriptor,
  buildProjectDescriptor,
  buildCommitDescriptor,
  buildBranchDescriptor,
  buildFileDescriptor,
  buildLayerDescriptor,
  buildCollectionDescriptor,
  buildActivityDescriptor,
  buildNotificationDescriptor,
  buildCommentDescriptor
} from "../support/factories";
import AbstractAPI from "./";

jest.mock("./randomTraceId");
jest.mock("../../package.json", () => ({
  ...jest.requireActual("../../package.json"),
  version: "0.0" // Mock version to make snapshots stable
}));

global.fetch = fetch;

const responses = {
  activities: {
    list: () => [
      JSON.stringify({
        data: {
          activities: [{ id: "foo" }, { id: "bar" }]
        }
      }),
      { status: 200 }
    ],
    info: () => [JSON.stringify({ id: "foo" }), { status: 200 }]
  },
  collections: {
    list: () => [
      JSON.stringify({
        data: {
          collections: [{ id: "first-collection" }, { id: "second-collection" }]
        }
      }),
      { status: 200 }
    ],
    info: () => [
      JSON.stringify({ collections: [{ id: "first-collection" }] }),
      { status: 200 }
    ]
  },
  branches: {
    list: () => [
      JSON.stringify({ data: [{ name: "branch-name" }] }),
      { status: 200 }
    ],
    info: () => [JSON.stringify({ name: "branch-name" }), { status: 200 }]
  },
  commits: {
    list: () => [
      JSON.stringify({
        commits: [{ sha: "commit-sha" }, { sha: "next-commit-sha" }]
      }),
      { status: 200 }
    ],
    info: () => [
      JSON.stringify({ commits: [{ sha: "commit-sha" }] }),
      { status: 200 }
    ]
  },
  files: {
    list: () => [
      JSON.stringify({ files: [{ id: "file-id" }, { id: "not-file-id" }] }),
      { status: 200 }
    ]
  },
  layers: {
    info: () => [
      JSON.stringify({
        layer: { name: "layer-name" },
        page: { name: "page-name", id: "page-id" },
        file: { name: "file-name" }
      }),
      { status: 200 }
    ]
  },
  previews: {
    arrayBuffer: (
      // inlined to avoid multiple reads
      data = fs.readFileSync(
        path.resolve(__dirname, "../../fixtures/preview.png")
      )
    ) => [data, { status: 200 }]
  },
  notifications: {
    list: () => [
      JSON.stringify({
        data: [{ id: "foo" }, { id: "bar" }]
      }),
      { status: 200 }
    ],
    info: () => [JSON.stringify({ id: "foo" }), { status: 200 }]
  },
  comments: {
    list: () => [
      JSON.stringify({
        data: [{ id: "foo" }, { id: "bar" }]
      }),
      { status: 200 }
    ],
    info: () => [JSON.stringify({ id: "foo" }), { status: 200 }]
  }
};

describe("AbstractAPI", () => {
  describe("with mocked global.fetch", () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test.each([
      // activities
      [
        "activities.list",
        buildBranchDescriptor(),
        { responses: [responses.activities.list()] }
      ],
      [
        "activities.list",
        buildOrganizationDescriptor(),
        { responses: [responses.activities.list()] }
      ],
      [
        "activities.list",
        buildProjectDescriptor(),
        { responses: [responses.activities.list()] }
      ],
      [
        "activities.info",
        buildActivityDescriptor(),
        { responses: [responses.activities.info()] }
      ],
      // organizations
      ["organizations.list", undefined],
      // projects
      ["projects.list", buildOrganizationDescriptor()],
      ["projects.list", [undefined, { filter: "active" }]],
      ["projects.info", buildProjectDescriptor()],
      // collections
      [
        "collections.list",
        buildProjectDescriptor(),
        { responses: [responses.collections.list()] }
      ],
      [
        "collections.list",
        buildBranchDescriptor(),
        { responses: [responses.collections.list()] }
      ],
      [
        "collections.info",
        buildCollectionDescriptor(),
        { responses: [responses.collections.list()] }
      ],
      // comments
      [
        "comments.create",
        [buildLayerDescriptor(), { body: "Comment on layer" }],
        {
          responses: [responses.branches.info(), responses.layers.info()]
        }
      ],
      [
        "comments.create",
        [
          buildLayerDescriptor(),
          {
            body: "Comment on layer with annotation",
            annotation: { x: 1, y: 1, width: 1, height: 1 }
          }
        ],
        { responses: [responses.branches.info(), responses.layers.info()] }
      ],
      [
        "comments.create",
        [buildBranchDescriptor(), { body: "Comment on branch HEAD" }],
        { responses: [responses.branches.info()] }
      ],
      [
        "comments.create",
        [
          buildBranchDescriptor({ sha: "my-sha" }),
          { body: "Comment on branch at my-sha" }
        ],
        { responses: [responses.branches.info()] }
      ],
      [
        "comments.list",
        buildProjectDescriptor(),
        { responses: [responses.comments.list()] }
      ],
      [
        "comments.info",
        buildCommentDescriptor(),
        { responses: [responses.comments.info()] }
      ],
      // commits
      ["commits.list", buildBranchDescriptor()],
      ["commits.list", buildFileDescriptor()],
      ["commits.list", buildLayerDescriptor()],
      [
        "commits.info",
        buildBranchDescriptor({ sha: "commit-sha" }),
        {
          responses: [responses.commits.info()],
          result: { sha: "commit-sha" }
        }
      ],
      [
        "commits.info",
        buildFileDescriptor({ sha: "commit-sha" }),
        {
          responses: [responses.commits.info()],
          result: { sha: "commit-sha" }
        }
      ],
      [
        "commits.info",
        buildLayerDescriptor({ sha: "commit-sha" }),
        {
          responses: [responses.commits.info()],
          result: { sha: "commit-sha" }
        }
      ],
      // branches
      [
        "branches.list",
        buildProjectDescriptor(),
        { responses: [responses.branches.list()] }
      ],
      [
        "branches.list",
        [buildProjectDescriptor(), { filter: "mine" }],
        { responses: [responses.branches.list()] }
      ],
      ["branches.info", buildBranchDescriptor()],
      // files
      ["files.list", buildBranchDescriptor()],
      [
        "files.info",
        buildFileDescriptor({ fileId: "file-id" }),
        {
          responses: [responses.files.list()],
          result: { id: "file-id" }
        }
      ],
      // changesets
      ["changesets.info", buildCommitDescriptor()],
      // pages
      ["pages.list", buildFileDescriptor()],
      // layers
      ["layers.list", buildFileDescriptor()],
      [
        "layers.info",
        buildLayerDescriptor(),
        { responses: [responses.layers.info()] }
      ],
      // previews
      [
        "previews.info",
        buildLayerDescriptor({
          projectId: "project-id",
          sha: "layer-sha",
          fileId: "file-id",
          layerId: "layer-id"
        })
      ],
      [
        "previews.raw",
        buildLayerDescriptor(),
        { responses: [responses.previews.arrayBuffer()] }
      ],
      // data
      ["data.info", buildLayerDescriptor()],
      ["data.info", buildLayerDescriptor({ sha: "sha" })],
      [
        "notifications.list",
        buildOrganizationDescriptor(),
        { responses: [responses.notifications.list()] }
      ],
      [
        "notifications.info",
        buildNotificationDescriptor(),
        { responses: [responses.notifications.info()] }
      ]
    ])("%s(%p)", async (property, args, options = {}) => {
      args = Array.isArray(args) ? args : [args];

      const transport = new AbstractAPI(buildOptions());
      const transportMethod = get(transport, property).bind(transport);

      if (options.responses) {
        fetch.mockResponses(...options.responses);
      }

      fetch.mockResponseOnce("{}");

      const result = transportMethod(...args);
      await expect(await result).resolves;

      if (options.result) {
        expect(await result).toEqual(options.result);
      }

      expect({ fetch: fetch.mock.calls }).toMatchSnapshot();
    });
  });
});
