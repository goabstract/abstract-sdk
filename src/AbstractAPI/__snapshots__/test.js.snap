// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractAPI with mocked global.fetch activities.list({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/activities?branchId=branch-id&projectId=project-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch activities.list({"organizationId": "organization-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/activities?organizationId=organization-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch activities.list({"projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/activities?projectId=project-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch branches.info({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch branches.list([[Object], [Object]]) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/?filter=mine",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch branches.list({"projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch changesets.info({"branchId": "branch-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits/commit-sha/changeset",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch collections.info({"branchId": "branch-id", "collectionId": "collection-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/collections/collection-id?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch collections.list({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/collections?branch_id=branch-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch collections.list({"projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/collections?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch comments.create([[Object], [Object]]) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits/layer-sha/files/file-id/layers/layer-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
    Array [
      "https://api.goabstract.com/comments",
      Object {
        "body": "{\\"projectId\\":\\"project-id\\",\\"branchId\\":\\"branch-id\\",\\"commitSha\\":\\"layer-sha\\",\\"fileId\\":\\"file-id\\",\\"pageId\\":\\"page-id\\",\\"layerId\\":\\"layer-id\\",\\"body\\":\\"Comment on layer\\",\\"branchName\\":\\"branch-name\\",\\"fileName\\":\\"file-name\\",\\"pageName\\":\\"page-name\\",\\"layerName\\":\\"layer-name\\"}",
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
        "method": "POST",
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch comments.create([[Object], [Object]]) 2`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits/layer-sha/files/file-id/layers/layer-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
    Array [
      "https://api.goabstract.com/comments",
      Object {
        "body": "{\\"projectId\\":\\"project-id\\",\\"branchId\\":\\"branch-id\\",\\"commitSha\\":\\"layer-sha\\",\\"fileId\\":\\"file-id\\",\\"pageId\\":\\"page-id\\",\\"layerId\\":\\"layer-id\\",\\"body\\":\\"Comment on layer with annotation\\",\\"annotation\\":{\\"x\\":1,\\"y\\":1,\\"width\\":1,\\"height\\":1},\\"branchName\\":\\"branch-name\\",\\"fileName\\":\\"file-name\\",\\"pageName\\":\\"page-name\\",\\"layerName\\":\\"layer-name\\"}",
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
        "method": "POST",
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch comments.create([[Object], [Object]]) 3`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
    Array [
      "https://api.goabstract.com/comments",
      Object {
        "body": "{\\"projectId\\":\\"project-id\\",\\"branchId\\":\\"branch-id\\",\\"body\\":\\"Comment on branch HEAD\\",\\"branchName\\":\\"branch-name\\"}",
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
        "method": "POST",
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch comments.create([[Object], [Object]]) 4`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
    Array [
      "https://api.goabstract.com/comments",
      Object {
        "body": "{\\"projectId\\":\\"project-id\\",\\"branchId\\":\\"branch-id\\",\\"commitSha\\":\\"my-sha\\",\\"body\\":\\"Comment on branch at my-sha\\",\\"branchName\\":\\"branch-name\\"}",
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
        "method": "POST",
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch commits.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch commits.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch commits.info({"branchId": "branch-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch commits.list({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits?fileId=file-id&layerId=layer-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch commits.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits?fileId=file-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch commits.list({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch data.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits/layer-sha/files/file-id/layers/layer-id/data",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch data.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits/sha/files/file-id/layers/layer-id/data",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch files.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/files",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch files.list({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/files",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch layers.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/commits/layer-sha/files/file-id/layers/layer-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch layers.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/files/file-id/layers?",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch organizations.list(undefined) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/organizations",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch pages.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id/branches/branch-id/files/file-id/pages",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch previews.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "fetch": Array [],
}
`;

exports[`AbstractAPI with mocked global.fetch previews.raw({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://previews.goabstract.com/projects/project-id/commits/layer-sha/files/file-id/layers/layer-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": undefined,
          "Accept": undefined,
          "Authorization": "Bearer abstract-token",
          "Content-Type": undefined,
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch projects.info({"projectId": "project-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects/project-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch projects.list([undefined, [Object]]) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects?filter=active",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;

exports[`AbstractAPI with mocked global.fetch projects.list({"organizationId": "organization-id"}) 1`] = `
Object {
  "fetch": Array [
    Array [
      "https://api.goabstract.com/projects?organizationId=organization-id",
      Object {
        "headers": Object {
          "Abstract-Api-Version": "7",
          "Accept": "application/json",
          "Authorization": "Bearer abstract-token",
          "Content-Type": "application/json",
          "User-Agent": "Abstract SDK 0.0",
          "X-Amzn-Trace-Id": "random-trace-id",
        },
      },
    ],
  ],
}
`;
